### PR TITLE
Reintroduce velocity drag

### PR DESCRIPTION
**Pull request by rjfarmer**

### CODE REVIEW

**Comment by rjfarmer** [comment]
Looks like we have a photo fail https://testhub.mesastar.org/reintroduce_velocity_drag/commits/6044d04

File: star/private/hydro_energy.f90 [hunk 1]
```diff
@@ -272,7 +273,24 @@ subroutine setup_sources_and_others(ierr) ! sources_ad, others_ad
...
+            drag_energy = 0d0
+            s% FdotV_drag_energy(k) = 0
+            if (s% q(k) > s% min_q_for_drag .and. s% drag_coefficient > 0) then
+               v_00 = wrap_v_00(s,k)
+               drag_force = s% drag_coefficient*v_00/s% dt
+               drag_energy = 0.5d0*v_00*drag_force
+               s% FdotV_drag_energy(k) = drag_energy%val
+            ! drag energy for outer half-cell.   the 0.5d0 is for dm/2
+            end if
+            if (s% q(k+1) > s% min_q_for_drag .and. s% drag_coefficient > 0) then
```
[the remark this answers is not in the document]
Comment (Debraheem) [reply star/private/hydro_energy.f90:285]: @rjfarmer This begs a second question. Is there a reason why we separate the cell into two parts in hydro_energy.f90, but we only use the top cell face velocity in hydro_momentum.f90? If there was a source for error in energy conservation, could this be the reason? I can add a conditional for k==s%nz, but using v_00 and not splitting the cell in two could also alleviate this.

[the remark this answers is not in the document]
Comment (rjfarmer) [reply star/private/hydro_energy.f90:285]: I assume the difference is that the velocity is defined at cell faces while the energy of a cell is over the whole cell

File: star/private/hydro_momentum.f90
Comment (fxt44) [inline star/private/hydro_momentum.f90:1]: good catch!

**Comment by rjfarmer** [comment]
Okay so after fixing the center zone now we need to decide what we want to do with the surface zone. 

At the moment expected_non_HSE_term (which sets dvdt_drag) does not get called for k=1 unless ``use_momentum_outer_BC = .true.``. So we either need to set it so you can't use the drag terms unless ``use_momentum_outer_BC = .true.`` or we set s% dvdt_drag(k=1)=0 and turn off the drag in the outer zone.  Thoughts @Debraheem ?

**Comment by Debraheem** [comment]
Hmm, Is there a situation where 'use_momentum_outer_BC = .true.' can't be used or causes issues during canonical stellar evolution? I have only used this boundary condition with Riemann hydro.

One example I could think of is on the surface of a post-agb wd, burning at the surface could drive the k= 1 zone crazy, unless drag is also applied in this zone. I think it's best to make the surface consistent with the layers below. 

We could do both: set the drag to 0 in the k = 1 zone if 'use_momentum_outer_BC = .false.', and apply the drag if 'use_momentum_outer_BC = .true.', then add a disclaimer in the documentation.

What do you think?

**Comment by Debraheem** [comment]
thanks @rjfarmer

**Comment by rjfarmer** [comment]
Well this is odd, we have a photo failure but no FPE's https://testhub.mesastar.org/reintroduce_velocity_drag/commits/5dc1a83/test_cases/star/make_pre_ccsn_13bvn

**Comment by evbauer** [comment]
That seems strange enough that I'll have Cannon run it again just to make sure.

**Comment by rjfarmer** [comment]
https://testhub.mesastar.org/reintroduce_velocity_drag/commits/0cb3533

make_pre_ccsn_13bvn is now passing. Unsurprisingly, it turns out that if you do not have a ck script then the ck command fails. Which looks like a photo checksum error.

Will merge once the last few tests finish

**Pull request merged**